rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -68,16 +71,25 @@ class MessagePoolMemoryStrategy
    */
   std::shared_ptr<MessageT> borrow_message()
   {
-    size_t current_index = next_array_index_;
-    next_array_index_ = (next_array_index_ + 1) % Size;
-    if (pool_[current_index].used) {
-      throw std::runtime_error("Tried to access message that was still in use! Abort.");
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+    if (free_list_.size() == 0) {
+      for (size_t i = 0; i < Size; ++i) {
+        if (pool_[i].use_count() == 1) {
+          free_list_.push_back(i);
+          break;
+        }
+      }
+      if (free_list_.size() == 0) {
+        throw std::runtime_error("No more free slots in the pool!");
```

Comment:
It would be nice if this was an specialized exception that could be captured for fixing when appears